src/persist/src/nemesis/validator.rs

```diff
@@ -0,0 +1,150 @@
+        }
+    }
+
+    fn check_success<T: fmt::Debug>(
+        &mut self,
+        req_id: ReqId,
+        res: &Result<T, Error>,
+        should_succeed: bool,
+    ) {
+        match (should_succeed, res) {
+            (true, Ok(_)) | (false, Err(_)) => {}
+            (true, Err(err)) => self.errors.push(format!(
+                "expected success but got error {:?}: {}",
+                req_id, err
+            )),
+            (false, Ok(res)) => self.errors.push(format!(
+                "expected error but got success {:?}: {:?}",
+                req_id, res
+            )),
+        }
+    }
+
+    fn step_write(&mut self, req_id: ReqId, req: WriteReq, res: Result<WriteRes, Error>) {
+        let should_succeed =
+            req.update.1 >= self.sealed.get(&req.stream).copied().unwrap_or_default();
+        self.check_success(req_id, &res, should_succeed);
+        if let Ok(res) = res {
+            self.writes_by_seqno
+                .insert((req.stream, SeqNo(res.seqno)), req.update);
+        }
+    }
+
+    fn step_seal(&mut self, req_id: ReqId, req: SealReq, res: Result<(), Error>) {
+        let should_succeed = req.ts > self.sealed.get(&req.stream).copied().unwrap_or_default();
+        self.check_success(req_id, &res, should_succeed);
+        if let Ok(_) = res {
+            self.sealed.insert(req.stream, req.ts);
+        }
+    }
+
+    fn step_take_snapshot(&mut self, req_id: ReqId, req: TakeSnapshotReq, res: Result<(), Error>) {
+        // At the moment, there are no nemeses that would prevent a snapshot
+        // from working, so we shouldn't see any errors.
+        let should_succeed = true;
+        self.check_success(req_id, &res, should_succeed);
+        if let Ok(_) = res {
+            self.available_snapshots.insert(req.snap, req.stream);
+        }
+    }
+
+    fn step_read_snapshot(
+        &mut self,
+        req_id: ReqId,
+        req: ReadSnapshotReq,
+        res: Result<ReadSnapshotRes, Error>,
+    ) {
+        match self.available_snapshots.remove(&req.snap) {
+            None => {
+                self.check_success(req_id, &res, false);
+            }
+            Some(stream) => {
+                self.check_success(req_id, &res, true);
+                if let Ok(res) = res {
+                    let mut actual = res.contents;
+                    actual.sort();
+                    let mut expected: Vec<((String, ()), u64, isize)> = self
+                        .writes_by_seqno
+                        .range((stream.to_owned(), SeqNo(0))..(stream.to_owned(), SeqNo(res.seqno)))
+                        .map(|(_, v)| v)
+                        .cloned()
+                        .collect();
+                    expected.sort();
+                    // TODO: We need to consolidate and account for since here once
+                    // that's supported in persist.
+                    if actual != expected {
```

Comment:
Once straight equality of actual v.s. expected becomes untenable, can you direct the actual and the expected to two new empty materialized arrangements and then read from them to obtain two comparable resultsets?

Reply:
Hmmm great question! Historically, I've tried to stress edge cases by writing test variants with different generator configurations that target the interactions I'm worried about instead of running things for a long time. You're right that, in this case, internal block sizes etc are going to make it more necessary to stress large data sizes. Perhaps if we make all those tunable, we can keep data sizes small enough that the Validator can deal with an entire history in memory?

If not, you're correct that an arrangement would be a great way to check this. However, there's actually a possibility that we'll end up using some of the differential dataflow trace stuff internally, which could mean we might lose the opportunity to catch a bug that affected both the real implementation and the verifier. A hash is also an interesting idea but I don't immediately see how to make the negations cancel out.

My current guess is that we'll push on this as much as possible by tweaking internal parameters around byte sizes and leave large data sizes for a separate test designed specifically for that, but I'm open to other opinions. Thoughts?

Reply:
Understood. When I said "untenable", I did not mean running out of memory, I mostly meant situations where, due to consolidation or another future operation/optimization, the "actual" ends up with a -2 diff while the "expected" had two -1 diffs or the timestamps get rewritten, that sort of suff.

Reply:
Oh! Yeah, for that case, I'd rather reimplement the consolidation logic in an obvious (and possibly slower) way inside validate and then call it on actual and expected before comparing them. It's nice for validate's code to be independent of what it's testing